lib/clients/public.js

```diff
@@ -1,206 +1,225 @@
+  get(...args) {
+    return this.request('get', ...args);
+  }
+  put(...args) {
+    return this.request('put', ...args);
+  }
+  post(...args) {
+    return this.request('post', ...args);
+  }
+  delete(...args) {
+    return this.request('delete', ...args);
+  }
 
-  prototype.addHeaders = function(obj, additional) {
+  addHeaders(obj, additional) {
     obj.headers = obj.headers || {};
-    return _.assign(obj.headers, {
-      'User-Agent': 'gdax-node-client',
-      'Accept': 'application/json',
-      'Content-Type': 'application/json',
-    }, additional);
-  };
-
-  prototype.makeRelativeURI = function(parts) {
+    return Object.assign(
+      obj.headers,
+      {
+        'User-Agent': 'gdax-node-client',
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      additional
+    );
+  }
+
+  makeRelativeURI(parts) {
     return '/' + parts.join('/');
-  };
+  }
 
-  prototype.makeAbsoluteURI = function(relativeURI) {
+  makeAbsoluteURI(relativeURI) {
     return this.apiURI + relativeURI;
-  };
+  }
 
-  prototype.makeRequestCallback = function(callback) {
-    return function(err, response, data) {
+  makeRequestCallback(callback, resolve, reject) {
+    return (err, response, data) => {
       try {
         data = JSON.parse(data);
       } catch (e) {
-        data = null
+        data = null;
+      }
+      if (typeof callback === 'function') {
+        callback(err, response, data);
+        return;
+      }
+      if (err) {
+        reject(err);
+      }
+      if (data === null) {
+        reject(new Error('Response could not be parsed as JSON'));
+      } else {
+        resolve(data);
       }
-      callback(err, response, data);
     };
-  };
+  }
 
-  prototype.request = function(method, uriParts, opts, callback) {
-    var self = this;
-    opts = opts || {};
-    if (!callback && (typeof opts === 'function')) {
+  request(method, uriParts, opts = {}, callback) {
+    if (!callback && typeof opts === 'function') {
       callback = opts;
       opts = {};
     }
-    if (!callback) {
-      throw "Must supply a callback."
-    }
-    _.assign(opts, {
-      'method': method.toUpperCase(),
-      'uri': self.makeAbsoluteURI(self.makeRelativeURI(uriParts)),
-    });
-    self.addHeaders(opts);
-    request(opts, self.makeRequestCallback(callback));
-  };
 
-  _.forEach(['get', 'post', 'put', 'delete'], function(method) {
-    prototype[method] = _.partial(prototype.request, method);
-  });
+    Object.assign(opts, {
+      method: method.toUpperCase(),
+      uri: this.makeAbsoluteURI(this.makeRelativeURI(uriParts)),
+    });
+    this.addHeaders(opts);
+    const p = new Promise((resolve, reject) => {
+      request(opts, this.makeRequestCallback(callback, resolve, reject));
+    });
 
-  prototype.getProducts = function(callback) {
-    var self = this;
-    return prototype.get.call(self, ['products'], callback);
-  };
+    if (callback) {
+      p.catch(() => {});
```

Comment:
This is definitely sensible behaviour, although I'd simply add listeners to the promise in case there is a callback. Still return the original promise (not the one with the callbacks applied) so that nothing needs to be returned from the callback.

We could also get rid of `makeRequestCallback` in favour of superagent promises — how do you feel about that?